controller-server/src/main/java/com/yahoo/vespa/hosted/controller/maintenance/SystemUpgrader.java

```diff
@@ -63,8 +63,14 @@ protected boolean shouldUpgrade(Version target, SystemApplication application, Z
             // For applications with package we do not have a zone-wide version target. This means that we must check
             // the wanted version of each node.
             return minVersion(zone, application, Node::wantedVersion)
-                    .map(target::isAfter)                                     // Upgrade if target is after any wanted version
-                    .orElse(true);                                            // Upgrade if there are no nodes allocated
+                    // Upgrade if target is after any wanted version
+                    .map(target::isAfter)
+                    // Skip upgrade if there are no nodes allocated. This is overloaded to mean that the zone is not
+                    // expected to have a deployment of this application.
+                    // TODO(mpolden): Once all zones are either directly routed or not: Change this to
+                    //                always deploy proxy app and wait for convergence in zones that are not directly
+                    //                routed.
+                    .orElse(false);
```

Comment:
I guess this would make it impossible to set up a new zone with a routing app, unless this was manually deployed first to give the application some nodes ... ? But that probably doesn't matter.

Reply:
Would probably need a manual deployment first yes, but it's the same behaviour as before.